// vhacd-js/lib/vhacd-wasm-api.js
/** Type declarations for the APIs provided by vhacd-wasm.js.
 * Those APIs are meant primarily for internal use; see vhacd.ts for the main JS API.
 */
export var VHACD;
(function (VHACD) {
    /** See HullFillMode in vhacd.ts. */
    let FillMode;
    (function (FillMode) {
        FillMode[FillMode["Flood"] = 0] = "Flood";
        FillMode[FillMode["Surface"] = 1] = "Surface";
        FillMode[FillMode["Raycast"] = 2] = "Raycast";
    })(FillMode = VHACD.FillMode || (VHACD.FillMode = {}));
    /** Specifies what kinds of messages generated by V-HACD to output to the console. */
    let MessageType;
    (function (MessageType) {
        MessageType[MessageType["None"] = 0] = "None";
        /** Indicates the stage and operation currently being performed. */
        MessageType[MessageType["Progress"] = 1] = "Progress";
        /** Informational and warning messages. */
        MessageType[MessageType["Log"] = 2] = "Log";
        MessageType[MessageType["All"] = 3] = "All";
    })(MessageType = VHACD.MessageType || (VHACD.MessageType = {}));
})(VHACD || (VHACD = {}));
